cardano-constitution/src/Cardano/Constitution/Validator/Dynamic.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE TemplateHaskell   #-}
+-- Following is for tx compilation
+{-# LANGUAGE DataKinds         #-}
+{-# OPTIONS_GHC -fplugin PlutusTx.Plugin #-}
+{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:target-version=1.1.0 #-}
+module Cardano.Constitution.Validator.Dynamic
+    ( constitutionValidator
+    , constitutionValidatorUntyped
+    , mkConstitutionCode
+    , defaultConstitutionValidator
+    , defaultConstitutionValidatorUntyped
+    , defaultConstitutionCode
+    ) where
+
+import Cardano.Constitution.Config
+import Cardano.Constitution.Predicate
+import Cardano.Constitution.ProtocolParam
+import Cardano.Constitution.Validator.Common
+import Cardano.Constitution.Validator.Dynamic.TH
+import PlutusCore.Version (plcVersion110)
+import PlutusLedgerApi.V3 as V3 hiding (ParamName)
+import PlutusTx as Tx
+import PlutusTx.AssocMap as Tx
+import PlutusTx.Prelude as Tx
+
+predicatesMeanings :: Tx.Map PredName PredicateMeaning
+predicatesMeanings = $$(resolvePredicatesMeanings)
+
+lookupMeaningDynamic :: PredName -> PredicateMeaning
+lookupMeaningDynamic predName = case Tx.lookup predName predicatesMeanings of
+                           Nothing -> Tx.error ()
+                           Just m  -> m
+
+-- | Check if a parameter's proposed new value satisfies the constitution config.
+satisfies :: ParamName -> PredName -> ConstitutionConfig -> Tx.Map ParamId PredValue -> Bool
+satisfies paramName predName cCfg cparams
+    | Just expectedPid <- fromParamName paramName cCfg
+    , Just expectedPredVal <- lookupPredValue paramName predName cCfg =
+            case Tx.lookup expectedPid cparams of
+                Just actualPredVal -> (lookupMeaningDynamic predName) expectedPredVal actualPredVal
+                -- This means that the expected parameter remains unchanged *OR*
+                -- the supplied index is unknown
+                Nothing            -> True
+    | otherwise = False
+
+-- Expects a constitution-configuration, statically *OR* at runtime via Tx.liftCode
+constitutionValidator :: ConstitutionConfig -> () -> V3.ScriptContext -> Bool
+constitutionValidator cCfg () =
+    \case
+        -- NOTE: this script will ignore (and thus succeed) on unknown indices inside cparams;
+        -- is it then the responsibility of the ledger to first sanitize cparams?
+        -- NOTE: this script will succeed on an empty cparams;
+        -- is it then the responsibility of the cardano-ledger to first sanitize cparams?
+        GetChangedParamsMap cparams ->
+            -- I could use `Tx.Bool.and` here, but probably is more costly
+                  (satisfies TxFeePerByte MinValue cCfg) cparams
```

Comment:
I don't think `TxFeePerByte` (and others) should be hardcoded here. You can extract a list of `[(ParamId, NewValue)]` from `ChangedParameters`, and for each `ParamId`, verify that the new value is within the valid range.

Reply:
Also, it doesn't seem like the use of `PredName` buys us anything, compared to the much simpler `newValue <= ub && newValue >= lb`.

Reply:
Outdated. Take a look at Unsorted.